### PR TITLE
tests(smoke): upload smokehouse artifacts on failure

### PR DESCRIPTION
**Pull request by connorjclark**
It'll be nice to be able to download the lhr / artifacts when smoke tests fail in CI.

Because not all smoke runs are done through the CLI runner, we're forced to rewrite the artifacts in `smokehouse.js`. It's not done in a way that makes `-G` work, but I think just being able to inspect the artifacts is enough for debugging purposes.

image

### CODE REVIEW

File: lighthouse-cli/test/smokehouse/test-definitions/seo/expectations.js [hunk 1]
```diff
@@ -142,7 +142,7 @@ const passing = {
         score: 1,
       },
       'document-title': {
-        score: 1,
+        score: 10,
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/test-definitions/seo/expectations.js:145]: TODO: undo, just testing

File: lighthouse-cli/test/smokehouse/lighthouse-runners/cli.js [hunk 2]
```diff
@@ -31,9 +32,10 @@ const ChildProcessError = require('../lib/child-process-error.js');
  * @return {Promise<{lhr: LH.Result, artifacts: LH.Artifacts, log: string}>}
  */
 async function runLighthouse(url, configJson, testRunnerOptions = {}) {
-  const tmpPath = await fs.mkdtemp(`${os.tmpdir()}/smokehouse-`);
-
   const {isDebug} = testRunnerOptions;
+  const tmpDir = isDebug ? `${LH_ROOT}/.tmp/smokehouse` : os.tmpdir();
+  await fs.mkdir(tmpDir, {recursive: true});
+  const tmpPath = await fs.mkdtemp(`${tmpDir}/smokehouse-`);
   return internalRun(url, tmpPath, configJson, testRunnerOptions)
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/lighthouse-runners/cli.js:39]: drive-by improvement: if debug mode, save the things in a readily-accessible folder inside `.tmp`.

Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/lighthouse-runners/cli.js:39]: > drive-by improvement: if debug mode, save the things in a readily-accessible folder inside .tmp.

good idea, though we delete the directory if `!isDebug`, so might as well just always put it in `.tmp` (I think we just hadn't settled on `.tmp` way back when and it felt gross just saving in root or whatever)

File: lighthouse-cli/test/smokehouse/smokehouse.js [hunk 3]
```diff
@@ -147,6 +149,18 @@ async function runSmokeTest(smokeTestDefn, testOptions) {
     report = getAssertionReport(result, expectations, {isDebug});
     if (report.failed) {
       bufferedConsole.log(`  ${getAssertionLog(report.failed)} failed.`);
+
+      // For CI, save to directory to be uploaded.
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/smokehouse.js:153]: what do you think about putting this in `smokehouse-bin.js` so `smokehouse.js` doesn't need to know anything about the filesystem (even if always guarded by `process.env.CI`)? The result/log just needs to be added to `SmokehouseResult`.

Comment (connorjclark) [reply lighthouse-cli/test/smokehouse/smokehouse.js:153]: each failed `report` would need to be tracked, and all of them returned as part of SmokehouseResult. seems like a larger change than you'd intend to make here?

Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/smokehouse.js:153]: that seems ok to me. It's just an array of stuff that's being passed back to `runSmokeTest` anyways.

File: lighthouse-cli/test/smokehouse/frontends/smokehouse-bin.js [hunk 4]
```diff
@@ -213,6 +215,24 @@ async function begin() {
 
   if (!smokehouseResult.success) {
     const failedTestResults = smokehouseResult.testResults.filter(r => r.failed);
+
+    // For CI, save to directory to be uploaded.
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/frontends/smokehouse-bin.js:219]:
```suggestion
    // For CI, save failed runs to directory to be uploaded.
```
maybe?

**Review by brendankenny — APPROVED** [review]
LGTM! I tried out the artifacts from one of the failed runs and these are really nice for debugging.
